microservices-orders/docs/index.rst

```diff
@@ -0,0 +1,578 @@
+      cp exercises/OrdersService.java kafka-streams-examples/src/main/java/io/confluent/examples/streams/microservices/.
+
+      # Compile the project and resolve any compilation errors
+      mvn clean compile -DskipTests -f kafka-streams-examples/pom.xml
+
+      # Run the test
+      mvn compile -Dtest=io.confluent.examples.streams.microservices.OrdersServiceTest test -f kafka-streams-examples/pom.xml
+
+
+Exercise 2: Event-driven App
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Service-based architectures are often designed to be request-driven, which sends commands to other services to tell them what to do, awaits a response, or sends queries to get the resulting state.
+
+.. figure:: images/event-driven.png
+    :alt: image
+
+    A visual summary of commands, events, and queries (`source <https://www.confluent.io/designing-event-driven-systems>`__)
+
+In contrast, in an event-driven design, there an event stream is the inter-service communication which leads to less coupling and queries, enables services to cross deployment boundaries, and avoids synchronous execution.
+
+.. figure:: images/microservices-exercise-2.jpg
+    :alt: image
+
+In this exercise, you will write a service that validates customer orders.
+Instead of using a series of synchronous calls to submit and validate orders, let the order event itself trigger the `OrderDetailsService`.
+When a new order is created, it is written to the topic `orders`, from which `OrderDetailsService` has a consumer polling for new records. 
+
+Implement the `TODO` lines of the file `exercises/OrderDetailsService.java <https://github.com/confluentinc/examples/tree/5.0.1-post/microservices-orders/exercises/OrderDetailsService.java>`__
+
+#. TODO 2.1: subscribe the existing `consumer` to a `Collections#singletonList` with the orders topic whose name is specified by `Topics.ORDERS.name()`
+#. TODO 2.2: validate the order using `OrderDetailsService#isValid` and save the validation result to type `OrderValidationResult`
+#. TODO 2.3: create a new record using `OrderDetailsService#result()` that takes the order and validation result
+#. TODO 2.4: produce the newly created record using the existing `producer`
+
+.. tip::
+
+   The following APIs will be helpful:
+
+   * https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/producer/KafkaProducer.html#send-org.apache.kafka.clients.producer.ProducerRecord-
+   * https://kafka.apache.org/20/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#subscribe-java.util.Collection-
+   * https://github.com/confluentinc/kafka-streams-examples/blob/5.0.1-post/src/main/java/io/confluent/examples/streams/microservices/domain/Schemas.java
+   * https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#singletonList-T-
+   
+   If you get stuck, here is the `complete solution <https://github.com/confluentinc/kafka-streams-examples/blob/5.0.1-post/src/main/java/io/confluent/examples/streams/microservices/OrderDetailsService.java>`__.
```

Comment:
This is also a meta question: I saw we use streams for some of the services but use producer / consumer for some other services (like this one), what's the motivation of doing that?

Reply:
@guozhangwang : It's intentionally showing REST API usage to produce events in the `orders` topic.  Original code is from @benstopford

Reply:
Ack, thanks for clarification (arguably we can also use Streams for REST API but I won't play the paranoid PITA here :P)